examples/src/main/java/org/apache/spark/examples/mllib/JavaSVDExample.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.examples.mllib;
+
+//$example on$
+import java.util.LinkedList;
+
+import org.apache.spark.SparkConf;
+import org.apache.spark.SparkContext;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.mllib.linalg.Matrix;
+import org.apache.spark.mllib.linalg.SingularValueDecomposition;
+import org.apache.spark.mllib.linalg.Vector;
+import org.apache.spark.mllib.linalg.Vectors;
+import org.apache.spark.mllib.linalg.distributed.RowMatrix;
+//$example off$
+
+/**
+ * Example for SingularValueDecomposition.
+ */
+public class JavaSVDExample {
+  public static void main(String[] args) {
+    SparkConf conf = new SparkConf().setAppName("SVD Example");
+    SparkContext sc = new SparkContext(conf);
+
+    // $example on$
+    double[][] array = { { 1.12, 2.05, 3.12 }, { 5.56, 6.28, 8.94 }, { 10.2, 8.0, 20.5 } };
+    LinkedList<Vector> rowsList = new LinkedList<Vector>();
+    for (int i = 0; i < array.length; i++) {
+      Vector currentRow = Vectors.dense(array[i]);
+      rowsList.add(currentRow);
+    }
+    JavaRDD<Vector> rows = JavaSparkContext.fromSparkContext(sc).parallelize(rowsList);
+
+    // Create a RowMatrix from JavaRDD<Vector>.
+    RowMatrix mat = new RowMatrix(rows.rdd());
+
+    // Compute the top 3 singular values and corresponding singular vectors.
+    SingularValueDecomposition<RowMatrix, Matrix> svd = mat.computeSVD(3, true, 1.0E-9d);
+    RowMatrix U = svd.U();
+    Vector s = svd.s();
+    Matrix V = svd.V();
+    Vector[] collectPartitions = (Vector[]) U.rows().collect();
```

Comment:
Ok let's keep it.

2016年2月17日星期三，Devaraj Kavali <notifications@github.com> 写道：

> In
> examples/src/main/java/org/apache/spark/examples/mllib/JavaSVDExample.java
> <https://github.com/apache/spark/pull/11132#discussion_r53277719>:
>
> > +    LinkedList<Vector> rowsList = new LinkedList<Vector>();
> > +    for (int i = 0; i < array.length; i++) {
> > +      Vector currentRow = Vectors.dense(array[i]);
> > +      rowsList.add(currentRow);
> > +    }
> > +    JavaRDD<Vector> rows = JavaSparkContext.fromSparkContext(sc).parallelize(rowsList);
> > +
> > +    // Create a RowMatrix from JavaRDD<Vector>.
> > +    RowMatrix mat = new RowMatrix(rows.rdd());
> > +
> > +    // Compute the top 3 singular values and corresponding singular vectors.
> > +    SingularValueDecomposition<RowMatrix, Matrix> svd = mat.computeSVD(3, true, 1.0E-9d);
> > +    RowMatrix U = svd.U();
> > +    Vector s = svd.s();
> > +    Matrix V = svd.V();
> > +    Vector[] collectPartitions = (Vector[]) U.rows().collect();
>
> It gives compilation error if we remove typecasting here since the return
> type of collect is Object.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/11132/files#r53277719>.
>


-- 
Cheers
-----------------------------------
Xusen Yin    (尹绪森)
LinkedIn: https://cn.linkedin.com/in/xusenyin